### PR TITLE
Fetcher syncer

### PR DESCRIPTION
**Pull request by Dcallies**
Summary
---------

Now connect to s3 and store state there as though they are files.

Test Plan
---------

Didn't test end-to-end, only by running locally and using the clowny path in the file.

```
$ python  -m hmalib.lambdas.fetcher
INFO:__main__:Triggered at time 15:55:58, found 2 collabs: HMA Test Collab 1, HMA Test Collab 2
INFO:__main__:Processing updates for collaboration HMA Test Collab 1
WARNING:__main__:No s3 checkpoint for 303636684709969. First run?
WARNING:__main__:Store for HMA Test Collab 1 - 303636684709969 stale! Resetting.
INFO:root:Fetch complete, applying 138 updates
INFO:__main__:138 rows stored for 303636684709969
INFO:__main__:Stored checkpoint for 303636684709969. last_fetch_time=1616108159 fetch_checkpoint=1616108158
INFO:__main__:Processing updates for collaboration HMA Test Collab 2
WARNING:__main__:No s3 checkpoint for 258601789084078. First run?
WARNING:__main__:Store for HMA Test Collab 2 - 258601789084078 stale! Resetting.
INFO:root:Fetch complete, applying 77 updates
INFO:__main__:77 rows stored for 258601789084078
INFO:__main__:Stored checkpoint for 258601789084078. last_fetch_time=1616108160 fetch_checkpoint=1616108160

$ python  -m hmalib.lambdas.fetcher
INFO:__main__:Triggered at time 16:25:53, found 2 collabs: HMA Test Collab 1, HMA Test Collab 2
INFO:__main__:Processing updates for collaboration HMA Test Collab 1
INFO:__main__:Loaded checkpoint for 303636684709969. last_fetch_time=1616108159 fetch_checkpoint=1616108158
INFO:root:Fetch complete, applying 0 updates
INFO:__main__:Stored checkpoint for 303636684709969. last_fetch_time=1616109954 fetch_checkpoint=1616109954
INFO:__main__:Processing updates for collaboration HMA Test Collab 2
INFO:__main__:Loaded checkpoint for 258601789084078. last_fetch_time=1616108160 fetch_checkpoint=1616108160
INFO:root:Fetch complete, applying 0 updates
INFO:__main__:Stored checkpoint for 258601789084078. last_fetch_time=1616109954 fetch_checkpoint=1616109954
```

### CODE REVIEW

File: python-threatexchange/threatexchange/threat_updates.py [hunk 1]
```diff
@@ -283,11 +287,10 @@ class ThreatUpdatesStore:
     def __init__(
         self,
         privacy_group: int,
-        types: t.Iterable[str] = (),
+        # TODO - remove: causes correctness issues with checkpoint
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/threat_updates.py:290]:
```suggestion
```

File: hasher-matcher-actioner/hmalib/lambdas/fetcher.py [hunk 2]
```diff
@@ -72,46 +92,218 @@ def lambda_handler(event, context):
...
+                logging.error("Failed before fetching any records")
 
     # TODO add TE data to indexer
 
     return {"statusCode": 200, "body": "Sure Yeah why not"}
 
+
+class ThreatUpdateS3PDQStore(tu.ThreatUpdatesStore):
+    """
+    Store files in S3!
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/fetcher.py:142]: nit; Store PDQ files in S3?

File: hasher-matcher-actioner/hmalib/lambdas/fetcher.py [hunk 3]
```diff
@@ -72,46 +92,218 @@ def lambda_handler(event, context):
...
+            items = []
+            if txt_content is None:
+                logger.warning("No TE state for %d. First run?", self.privacy_group)
+            else:
+                # Violate your warranty with module state!
+                csv.field_size_limit(65535)  # dodge field size problems
+                for row in csv.reader(txt_content):
+                    items.append(
+                        CliIndicatorSerialization(
+                            "HASH_PDQ",
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/fetcher.py:236]: Can we store "HASH_PDQ" as a class constant?

File: python-threatexchange/threatexchange/threat_updates.py (hunk 1 above)
Comment (jeberl) [inline python-threatexchange/threatexchange/threat_updates.py]: Did you mean to just comment out line 286?

**Review by jeberl — APPROVED** [review]
Looks great!

**Pull request merged**